src/seabreeze/cseabreeze/c_seabreeze_wrapper.pyx

```diff
@@ -841,14 +861,112 @@ cdef class SeaBreezeSpectrometerFeature(SeaBreezeFeature):
+            data = c_buffer[:bytes_written]
+        finally:
+            PyMem_Free(c_buffer)
+
+        # Define Dataset structure for individual measurements.
+        Dataset = namedtuple("Dataset", [
+                                        "metadata_protocol_version",
+                                        "metadata_length",
+                                        "pixel_data_length",
+                                        "microsecond_counter",
+                                        "integration_time_micros",
+                                        "pixel_data_format",
+                                        "spectrum_count",
+                                        "last_spectrum_count",
+                                        "last_microsecond_count",
+                                        "scans_to_average",
+                                        "intensities"
+                                        ])
+        buffer_data = []
+        offset = 0
+        for i in range(sample_number):
+            # decode metadata
+            protocol_version = np.frombuffer(data[offset+0:offset+2], dtype=np.uint16)[0]
+            metadata_length = np.frombuffer(data[offset+2:offset+4], dtype=np.uint16)[0]
+            pixel_data_length = np.frombuffer(data[offset+4:offset+8], dtype=np.uint32)[0]
+            microsecond_counter = np.frombuffer(data[offset+8:offset+16], dtype=np.uint64)[0]
+            integration_time_micros = np.frombuffer(data[offset+16:offset+20], dtype=np.uint32)[0]
+            pixel_data_format = np.frombuffer(data[offset+20:offset+24], dtype=np.uint32)[0]
+            spectrum_count = np.frombuffer(data[offset+24:offset+28], dtype=np.uint32)[0]
+            last_spectrum_count = np.frombuffer(data[offset+28:offset+32], dtype=np.uint32)[0]
+            last_microsecond_count = np.frombuffer(data[offset+32:offset+40], dtype=np.uint64)[0]
+            scans_to_average = np.frombuffer(data[offset+40:offset+42], dtype=np.uint16)[0]
+
+            # determine pixel data format from metadata and decode spectrum
+            if pixel_data_format == 1:
+                intensities = np.frombuffer(data[offset+metadata_length:offset+metadata_length+pixel_data_length], dtype=np.uint16)
+            elif pixel_data_format == 2:
+                intensities = np.frombuffer(data[offset+metadata_length:offset+metadata_length+pixel_data_length], dtype=np.uint24)
+            elif pixel_data_format == 3:
+                intensities = np.frombuffer(data[offset+metadata_length:offset+metadata_length+pixel_data_length], dtype=np.uint32)
+            elif pixel_data_format == 4:
+                intensities = np.frombuffer(data[offset+metadata_length:offset+metadata_length+pixel_data_length], dtype=np.single)
+            else:
+                SeaBreezeError("Unknown Pixel Data Format")
+
+            # add Dataset to returned list
+            buffer_data.append(Dataset(
```

Comment:
I would separate this. Create DatasetMetadata namedtuples and numpy arrays containing the intensities.

The return type could then be `list[tuple[DatasetMetadata, np.Array]]` or `tuple[list[DatasetMetadata], list[np.Array]]` whatever you think fits better

Reply:
I wasn't quite sure about how to implement this myself. I guess it would be nice to have all spectra in their own list. But then that would also be nice to have all timestamps in their own list. So in the end I opted for the first option as in `list[tuple[DatasetMetadata, np.Array]]`. This way I feel like the structure is a bit more consistent if not as easy to work with...